llvm/lib/Target/AIE/AIEPostPipeliner.cpp

```diff
@@ -33,6 +36,58 @@ class PostPipelineDumper : public PipelineScheduleVisitor {
   void endBundle() override { dbgs() << "\t}\n"; }
 };
 
+SlotCounts::SlotCounts(SlotBits Bits) {
+  while (Bits) {
+    assert(Size < MaxSlots);
+    Counts[Size] = Bits & 1;
+    Size++;
+    Bits >>= 1;
+  }
+}
+
+SlotCounts::SlotCounts(const SlotCounts &Org) : Size(Org.Size) {
+  for (int I = 0; I < Size; I++) {
+    Counts[I] = Org.Counts[I];
+  }
+}
+
+int SlotCounts::max() {
+  int Max = 0;
+  for (int I = 0; I < Size; I++) {
+    Max = std::max(Max, int(Counts[I]));
+  }
+  return Max;
+}
+
+SlotCounts &SlotCounts::operator+=(const SlotCounts &Other) {
+  // The common part
+  for (int I = 0; I < Size && I < Other.Size; I++) {
+    Counts[I] += Other.Counts[I];
+  }
+  // Any excess from the other
+  while (Size < Other.Size) {
+    Counts[Size++] = Other.Counts[Size];
+  }
+  assert(Size >= Other.Size);
+  assert(Size < MaxSlots);
+  return *this;
+}
+
+SlotCounts SlotCounts::operator+(const SlotCounts &Other) const {
+  SlotCounts Result(*this);
+  return Result += Other;
+}
+llvm::raw_ostream &operator<<(llvm::raw_ostream &OS, const SlotCounts &Val) {
+  OS << "{ ";
+  const char *Sep = "";
+  for (int I = 0; I < Val.size(); I++) {
+    OS << Sep << Val[I];
+    Sep = ", ";
+  }
+  OS << " }";
+  return OS;
+}
```

Comment:
I think this new `SlotCounts` should be lifted to its own header. This can be useful beyond the `MachinePipeliner`